File: smtk/attribute/Testing/expressionTest.cxx
/*=========================================================================

Copyright (c) 1998-2012 Kitware Inc. 28 Corporate Drive,
Clifton Park, NY, 12065, USA.

All rights reserved. No part of this software may be reproduced, distributed,
or modified, in any form or by any means, without permission in writing from
Kitware Inc.

IN NO EVENT SHALL THE AUTHORS OR DISTRIBUTORS BE LIABLE TO ANY PARTY FOR
DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING OUT
OF THE USE OF THIS SOFTWARE, ITS DOCUMENTATION, OR ANY DERIVATIVES THEREOF,
EVEN IF THE AUTHORS HAVE BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

THE AUTHORS AND DISTRIBUTORS SPECIFICALLY DISCLAIM ANY WARRANTIES,
INCLUDING,
BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
PARTICULAR PURPOSE, AND NON-INFRINGEMENT.  THIS SOFTWARE IS PROVIDED ON AN
"AS IS" BASIS, AND THE AUTHORS AND DISTRIBUTORS HAVE NO OBLIGATION TO
PROVIDE
MAINTENANCE, SUPPORT, UPDATES, ENHANCEMENTS, OR MODIFICATIONS.
=========================================================================*/

#include "smtk/attribute/Manager.h"
#include "smtk/attribute/AttributeDefinition.h"
#include "smtk/attribute/Attribute.h"
#include "smtk/attribute/IntItem.h"
#include "smtk/attribute/IntItemDefinition.h"
#include "smtk/attribute/DoubleItem.h"
#include "smtk/attribute/DoubleItemDefinition.h"
#include "smtk/attribute/StringItem.h"
#include "smtk/attribute/StringItemDefinition.h"
#include <iostream>

int main()
{
  int status = 0;
  {
  typedef smtk::attribute::IntItemDefinition IntItemDef;
  typedef smtk::attribute::DoubleItemDefinition DoubleItemDef;
  typedef smtk::attribute::StringItemDefinition StringItemDef;
  typedef smtk::attribute::ValueItem ValueItem;
  typedef smtk::attribute::Item AttItem;

  smtk::attribute::Manager manager;
  std::cout << "Manager Created\n";
  // Lets create an attribute to represent an expression
  smtk::AttributeDefinitionPtr expDef = manager.createAttributeDefinition("ExpDef");
  smtk::StringItemDefinitionPtr eitemdef = StringItemDef::New("Expression String");
  expDef->addItemDefinition(eitemdef);

  smtk::AttributeDefinitionPtr base = manager.createAttributeDefinition("BaseDef");
  // Lets add some item definitions
  smtk::IntItemDefinitionPtr iitemdef = IntItemDef::New("IntItem1");
  base->addItemDefinition(iitemdef);
  smtk::IntItemDefinitionPtr iitemdef2 = IntItemDef::New("IntItem2");
  iitemdef2->setDefaultValue(10);
  base->addItemDefinition(iitemdef2);

  smtk::AttributeDefinitionPtr def1 = manager.createAttributeDefinition("Derived1", "BaseDef");
   // Lets add some item definitions
  smtk::DoubleItemDefinitionPtr ditemdef = DoubleItemDef::New("DoubleItem1");
  // Allow this one to hold an expression
  ditemdef->setExpressionDefinition(expDef);
  // Check to make sure we can use expressions
  if (!ditemdef->allowsExpressions())
    {
    std::cout << "ERROR - Item Def does not allow expressions\n";
    status = -1;
    }
  def1->addItemDefinition(ditemdef);
  smtk::DoubleItemDefinitionPtr ditemdef2 = DoubleItemDef::New("DoubleItem2");
  ditemdef2->setDefaultValue(-35.2);
  def1->addItemDefinition(ditemdef2);

  smtk::AttributeDefinitionPtr def2 = manager.createAttributeDefinition("Derived2", "Derived1");
   // Lets add some item definitions
  smtk::StringItemDefinitionPtr sitemdef = StringItemDef::New("StringItem1");
  def1->addItemDefinition(sitemdef);
  smtk::StringItemDefinitionPtr sitemdef2 = StringItemDef::New("StringItem2");
  sitemdef2->setDefaultValue("Default");
  def1->addItemDefinition(sitemdef2);

  // Lets test creating an attribute by passing in the expression definition explicitly
  smtk::AttributePtr expAtt = manager.createAttribute("Exp1", expDef);
  smtk::AttributePtr att = manager.createAttribute("testAtt", "Derived2");
  if (att != NULL)
    {
    std::cout << "Attribute testAtt created\n";
    }
  else
    {
    std::cout << "ERROR: Attribute testAtt not created\n";
    status = -1;
    }

  smtk::ValueItemPtr vitem;
  smtk::AttributeItemPtr item;

  // Find the expression enabled item
  item = att->item(2);
  vitem = smtk::dynamicCastPointer<ValueItem>(item);
  if (vitem->allowsExpressions())
    {
    vitem->setExpression(expAtt);
    std::cout << "Expression Set on " << vitem->name() << "\n";
    }
  else
    {
    std::cout << "ERROR: Can not set expression on " << vitem->name() << "\n";
    status = -1;
    }

  int i, n = att->numberOfItems();
  std::cout << "Items of testAtt:\n";
  for (i = 0; i < n; i++)
    {
    item = att->item(i);
    std::cout << "\t" << item->name() << " Type = " << AttItem::type2String(item->type()) << ", ";
    vitem = smtk::dynamicCastPointer<ValueItem>(item);
    if (vitem != NULL)
      {
      if (vitem->isExpression())
        {
        std::cout << " using Expression: " << vitem->expression()->name() << "\n";
        }
      else
        {
        switch (vitem->type())
          {
          case AttItem::DOUBLE:
          case AttItem::INT:
            std::cout << " Value = "  << vitem->valueAsString() << std::endl;
            break;
          case AttItem::STRING:
            std::cout << " String Val = " << vitem->valueAsString() << std::endl;
            break;
          default:
            break;
          }
        }
      }
    }
  std::cout << "Manager destroyed\n";
  }
  return status;
}

File: smtk/attribute/testing/python/expressionTest.py
#=============================================================================
#
#  Copyright (c) Kitware, Inc.
#  All rights reserved.
#  See LICENSE.txt for details.
#
#  This software is distributed WITHOUT ANY WARRANTY; without even
#  the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
#  PURPOSE.  See the above copyright notice for more information.
#
#=============================================================================
"""
    Manual port of SMTK/smtk/attribute/Testing/expressionTest.cxx
    For verifying python-shiboken wrappers

    Requires smtkCorePython.so to be in module path
    """

import smtk
from smtk import attribute

if __name__ == '__main__':
    import sys
    import os

    status = 0

    system = smtk.attribute.System()
    print 'System created'
    # Lets create an attribute to represent an expression
    expDef = system.createDefinition("ExpDef")
    eitemdef = smtk.attribute.StringItemDefinition.New("Expression String")
    expDef.addItemDefinition(eitemdef)

    base = system.createDefinition("BaseDef")
    # Lets add some item definitions
    iitemdef = smtk.attribute.IntItemDefinition.New("IntItem1")
    base.addItemDefinition(iitemdef)
    iitemdef2 = smtk.attribute.IntItemDefinition.New("IntItem2")
    base.addItemDefinition(iitemdef2)
    iitemdef2.setDefaultValue(10)

    def1 = system.createDefinition("Derived1", "BaseDef")
    # Lets add some item definitions
    ditemdef = smtk.attribute.DoubleItemDefinition.New("DoubleItem1")
    def1.addItemDefinition(ditemdef)
    # Allow this one to hold an expression
    ditemdef.setExpressionDefinition(expDef)
    # Check to make sure we can use expressions
    if not ditemdef.allowsExpressions():
        print "ERROR - Item Def does not allow expressions"
        status = -1
    ditemdef2 = smtk.attribute.DoubleItemDefinition.New("DoubleItem2")
    def1.addItemDefinition(ditemdef2)
    ditemdef2.setDefaultValue(-35.2)

    def2 = system.createDefinition("Derived2", "Derived1")
    # Lets add some item definitions
    sitemdef = smtk.attribute.StringItemDefinition.New("StringItem1")
    def1.addItemDefinition(sitemdef)
    sitemdef2 = smtk.attribute.StringItemDefinition.New("StringItem2")
    def1.addItemDefinition(sitemdef2)
    sitemdef2.setDefaultValue("Default")

    # Lets test creating an attribute by passing in the expression definition
    # explicitly
    expAtt = system.createAttribute("Exp1", expDef)
    att = system.createAttribute("testAtt", "Derived2")
    if att is not None:
        print "Attribute testAtt created"
    else:
        print "ERROR: Attribute testAtt not created"
        status = -1

    # Find the expression enabled item
    vitem = smtk.attribute.ValueItem.CastTo(att.item(2))
    if vitem.allowsExpressions():
        vitem.setExpression(expAtt)
        print "Expression Set on %s" % vitem.name()
    else:
        print "ERROR: Can not set expression on %s" % vitem.name()
        status = -1

    n = att.numberOfItems()
    print "Items of testAtt:"
    for i in range(0, n, 1):
        item = att.item(i)
        sys.stdout.write("\t%s Type = %s, " % (
            item.name(), smtk.attribute.Item.type2String(item.type())))
        vitem = smtk.attribute.ValueItem.CastTo(item)
        if vitem is not None:
            if vitem.isExpression(0):
                print " using Expression: %s" % vitem.expression(0).name()
            else:
                t = vitem.type()
                if t == smtk.attribute.Item.DOUBLE or t == smtk.attribute.Item.INT:
                    print " Value = %s" % vitem.valueAsString()
                elif t == smtk.attribute.Item.STRING:
                    print " String Val = %s" % vitem.valueAsString()

    del system
    print 'System destroyed'

    sys.exit(status)
